### DISCUSSION TITLE
Allowing a class object to satisfy a protocol

### DISCUSSION THREAD

**Discussion by ronf** [python/typing#862] [Q&A]
I have a situation where I have a protocol I'd like to define which can be implemented either as instance methods in a normal object or in class methods in a class, when I don't need multiple instances. I could do this with a singleton, but Python has no problem letting me just pass in the class itself in place of an instance, as long as all the methods that would have been instance methods are made into class methods. Here's a trimmed down example showing the basic idea:

```python
from typing_extensions import Protocol

class MyProtocol(Protocol):
    def test(self, x: int) -> None: ...

class NormalClass:
    def test(self, x: int) -> None:
        print('Foo:', x)

class ClassAsInstance:
    @classmethod
    def test(cls, x: int) -> None:
        print('Bar:', x)

foo: MyProtocol = NormalClass()
bar: MyProtocol = ClassAsInstance

foo.test(10)
bar.test(20)
```

If you run this code, it works just fine. However, when running it through mypy, I get:

```
/tmp/t.py:16: error: Incompatible types in assignment (expression has type "Type[ClassAsInstance]", variable has type "MyProtocol")
```

Any thoughts on whether something like this is possible?

**Comment by gvanrossum** [comment]
That's not unreasonable. I think we once tried to make it possible to describe a module using a Protocol (can't recall if we managed that) and it could work for classes too. I have no idea how hard it would be to implement though.

**Reply by ronf** [reply]
Thanks for the confirmation. I also did some testing with static methods, and that also fails. Looking at the error message, I think it may be failing before it even starts to look at the methods defined in the class. I'm thinking that a "Type" object like this might never be considered compatible with a Protocol right now. Note also that it doesn't list any specific methods as being missing the way it normally does when you pass in an instance of a class which doesn't conform.

For now, I've gone ahead and changed my singleton class to use instance methods instead, replacing the class reference with a reference to a global singleton created at module load time. It was actually less disruptive than I originally thought it would be. However, since this pattern is something Python supports, I can see potential value in mypy understanding that all class methods of an object of type class are basically instance methods, and static methods on the class can also match static methods in the protocol.

**Reply by JelleZijlstra** [reply]
https://github.com/python/mypy/issues/4536 is the relevant issue.

**Reply by ronf** [reply]
Ah - indeed it is, thanks!
